### PR TITLE
Remove ListWatchUntil

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #90813

This isn't per-se improving situation of that bug, this `ListWatchUntil` seem to be an unnecessary function currently.

```release-note
NONE
```

/kind cleanup

### CODE REVIEW

**Comment by liggitt** [comment]
/assign @tnozicka 

I forget the pedigree of this method... do you remember the reason it was added?

**Comment by tnozicka** [comment]
I think the specific of ListWatchUntil is that you get guaranteed order and presence of all events in contrast to informer based UntilWithSync. While informer based approach watching the levels is what you'd use in production code, testing suites might want to check invariants which require being able to see all events and the order. I haven't given much though yet to if you'd mostly use a manual list and Until in such case yet, but If you'd feel strong about removing it I can do so next week. (Today is public holiday in Czech republic so I am out.)

**Comment by tnozicka** [comment]
https://github.com/kubernetes/kubernetes/blob/603dd25/staging/src/k8s.io/client-go/tools/watch/until.go#L99-L100

**Comment by wojtek-t** [comment]
> I think the specific of ListWatchUntil is that you get guaranteed order and presence of all events in contrast to informer based UntilWithSync. 

The order is guaranteed in informer too. It's just some events might be potentially compacted in informer.

That said - if we can re-watch from the point when last watch was finished, informer will also do that. And if we can't (too old RV, out-of-history), the ListWatchUntil will fail anyway. The informer will relist in that case.
Re setting exact starting point - I did that in this PR in this only place that uses it already.

With bookmarks currently being send every 1m: #90560 the informer should be pretty reliable.

So given the above (as well as the fact that ListWatchUntil underneath uses Until which we should get rid of), I actually think we should remove it.

@liggitt - WDYT?

**Comment by liggitt** [comment]
> It's just some events might be potentially compacted in informer.

That is not desired for some tests.

> if we can re-watch from the point when last watch was finished, informer will also do that. And if we can't (too old RV, out-of-history), the ListWatchUntil will fail anyway. The informer will relist in that case.

If the test needs to verify every watch event, silently masking the fact that a watch event was missed might not be what is desired.

I think `Until` still has a place for tests like that, especially given that we expect bookmark events to make it much more reliable

**Comment by wojtek-t** [comment]
> That is not desired for some tests.

OK - let me rephrase my arguments a bit.
Currently "ListWatchUntil" is used in exactly one place in the whole codebase.
And in this place, the fact that couple versions will be compacted doesn't matter, because what we only care about is we wait for "delete" event'. [Delete can be compacted only with further create, which is not the case in this particualr test].

So I think that maintaining this function for one case, where it isn't needed doesn't make sense.

@liggitt - does that make more sense?

**Comment by wojtek-t** [comment]
At the same time, I agree that we need "until" functionality, but I think the "Until" is enough.

@liggitt - PTAL

**Comment by liggitt** [comment]
> Currently "ListWatchUntil" is used in exactly one place in the whole codebase.
> And in this place, the fact that couple versions will be compacted doesn't matter, because what we only care about is we wait for "delete" event'.

ok

> At the same time, I agree that we need "until" functionality, but I think the "Until" is enough.

that sounds ok

**Comment by wojtek-t** [comment]
Unrelated flakes
/retest

**Comment by tnozicka** [comment]
>At the same time, I agree that we need "until" functionality, but I think the "Until" is enough.

The issue with `Until` or rather watches in general is that it can't start from `RV=""` and you have to make a list manually before to obtain a RV. `ListWatchUntil` was filling that space. I agree the usefulness is limited though.

I don't think internal usage is the only metric as this is part of client-go, on the other hand I understand the cost of maintaining it.

**Comment by Dean-Coakley** [comment]
> https://github.com/kubernetes/kubernetes/blob/603dd25/staging/src/k8s.io/client-go/tools/watch/until.go#L99-L100

@tnozicka @wojtek-t 
Does it make sense to simply remove?: `If you want the initial list ("", "0") done for you use ListWatchUntil instead.` 

Or is there something more reasonable to add in its place?

**Comment by tnozicka** [comment]
> Does it make sense to simply remove?: If you want the initial list ("", "0") done for you use ListWatchUntil instead.

I think simply removing it is fine.
